### PR TITLE
Minor Improvements

### PR DESCRIPTION
**Pull request by k8culver**
- Added "Show Conflicts" toggle
- Fixed resize bug
- Minor refactoring to clean up dash_html.py & app.py

### CODE REVIEW

File: dash_html.py [hunk 1]
```diff
@@ -38,14 +38,20 @@
...
+    THUMBNAIL,
 )
+from src.job_shop_scheduler import HybridSamplerType, SamplerType
 
-SOLVER_OPTIONS = ["Quantum Hybrid (CQM)", "Quantum Hybrid (NL)", "Classical (HiGHS)"]
+SAMPLER_TYPES = {
+    SamplerType.HYBRID: "Quantum Hybrid" if SHOW_CQM else "Quantum Hybrid (NL)",
+    SamplerType.HIGHS: "Classical (HiGHS)",
+}
+HYBRID_SAMPLER_TYPES = {HybridSamplerType.NL: "NL", HybridSamplerType.CQM: "CQM"}
```
Comment (thisac) [inline dash_html.py:54]: (minor) Aesthetically, I think `NL` and `CQM` looks a bit short in the dropdown. Perhaps use the full names here instead?
```suggestion
HYBRID_SAMPLER_TYPES = {HybridSamplerType.NL: "Nonlinear (NL)", HybridSamplerType.CQM: "Constraint Quadratic Model (CQM"}
```

File: src/job_shop_scheduler.py [hunk 2]
```diff
@@ -3,24 +3,36 @@
...
-from dwave.optimization.generators import flow_shop_scheduling
+
+class SamplerType(Enum):
+    HYBRID = 0
+    HIGHS = 1
+
+
+class HybridSamplerType(Enum):
+    CQM = 0
+    NL = 1
```
Comment (thisac) [inline src/job_shop_scheduler.py:35]: NL should be the default selected solver.
```suggestion
class HybridSamplerType(Enum):
    NL = 0
    CQM = 1
```

File: app.py [hunk 3]
```diff
@@ -127,53 +123,40 @@ def toggle_left_column(collapse_trigger: int, to_collapse_class: str) -> str:
...
     inputs=[
         Input("solver-select", "value"),
-        State("solver-select", "options"),
     ],
     prevent_initial_call=True,
 )
 def update_solvers_selected(
     selected_solvers: list[int],
-    solver_options: list[dict]
 ) -> list[dict]:
```
Comment (thisac) [inline app.py:134]:
```suggestion
) -> str:
```

File: app.py [hunk 4]
```diff
@@ -127,53 +123,40 @@ def toggle_left_column(collapse_trigger: int, to_collapse_class: str) -> str:
...
-
-    Note, can be removed if CQM solver is not in use.
+    """Hide NL/CQM selector when Hybrid is unselected. Not applicable when SHOW_CQM is False.
 
     Args:
         selected_solvers (list[int]): Currently selected solvers.
-        solver_options (list[dict]): List of solver checkbox options.
 
     Returns:
         list: Updated list of solver checkbox options.
```
Comment (thisac) [inline app.py:141]:
```suggestion
        str: Class name for hybrid select wrapper.
```

File: app.py [hunk 5]
```diff
@@ -268,130 +241,185 @@ def update_button_visibility(running_dwave: bool, running_classical: bool) -> tu
...
+    scenario: str, solver: str, model_data: JobShopData, time_limit: int, wall_clock_time: float
+) -> tuple[dict]:
+    """Returns tuple of dicts where each dict represents a row in the problem details table."""
+    return (
+        {"Scenario": scenario, "Solver": solver},
+        {"Number of Jobs": model_data.get_job_count(), "Solver Time Limit": f"{time_limit}s"},
+        {
+            "Number of Operations": model_data.get_resource_count(),
+            "Wall Clock Time": f"{round(wall_clock_time, 2)}s",
+        },
```
Comment (thisac) [inline app.py:290]: Why not just put this inside the `generate_problem_details_table` function instead?

File: app.py [hunk 6]
```diff
@@ -268,130 +241,185 @@ def update_button_visibility(running_dwave: bool, running_classical: bool) -> tu
...
+
+    gantt_chart_jobsort: go.Figure
+    gantt_chart_startsort: go.Figure
+    dwave_makespan: str
+    dwave_solution_stats_table: list
+    dwave_tab_disabled: bool
+    dwave_gantt_title_span: str
+    dwave_tab_class: str
+    dwave_tab_label: str
+    running_dwave: bool
```
Comment (thisac) [inline app.py:305]: We could set `dash.no_update` as defaults to all
```python
class RunOptimizationHybridReturn(NamedTuple):
    """Return type for the ``run_optimization_hybrid`` callback function."""

    gantt_chart_jobsort: go.Figure = dash.no_update
    gantt_chart_startsort: go.Figure = dash.no_update
    dwave_makespan: str = dash.no_update
    ...
```
and then just set the ones we need as kwargs
```python
return RunOptimizationHybridReturn(
    dwave_tab_class="tab", 
    dwave_tab_label=DWAVE_TAB_LABEL, 
    running_dwave=False
)
```
Same for the other named tuple returns.

File: dash_html.py [hunk 7]
```diff
@@ -124,64 +237,65 @@ def generate_control_card() -> html.Div:
...
+        row = []
+        for label, cell in table_row.items():
+            row += [label, cell]
+        rows.append(row)
+
+    return (
+        html.Tbody(
+            children=[html.Tr([html.Td(cell) for cell in row]) for row in rows],
+        ),
+    )
```
Comment (thisac) [inline dash_html.py:260]: Does this need to be returned as a tuple?

File: dash_html.py [hunk 8]
```diff
@@ -124,64 +237,65 @@ def generate_control_card() -> html.Div:
...
+    rows = []
+    for table_row in table_rows:
+        row = []
+        for label, cell in table_row.items():
+            row += [label, cell]
+        rows.append(row)
+
+    return (
+        html.Tbody(
+            children=[html.Tr([html.Td(cell) for cell in row]) for row in rows],
```
Comment (thisac) [inline dash_html.py:258]: Looks like this could be simplified one loop by creating the `html.Tr` and `html.Td` objects in the double-loop above.

**Review by thisac — COMMENTED** [review]
Thanks for the updates @k8culver! Just a few comments and thoughts.

* The legend now has a conflicts column (adding an _extra_ item for each "conflicted" colour) which I don't think is necessary. It looks better, and to me clearer, just to have the colours in the legend.
* Not sure if this is easily fixable, but it'd be nice to have the conflicts button switch between the two graphs as smoothly as the sort button. Can both images be rendered at start and just hidden in the same was as the sorted ones?

File: dash_html.py (hunk 7 above)
Comment (k8culver) [reply dash_html.py:260]: `black` did this and I have no idea why

File: dash_html.py (hunk 8 above)
Comment (k8culver) [reply dash_html.py]: Changed this all around so it's less redundant: https://github.com/dwave-examples/flow-shop-scheduling/pull/5/commits/6c9e8ca64844a1f71962ed15d87350bde9b8ee47

File: dash_html.py (hunk 7 above)
Comment (k8culver) [reply dash_html.py]: updated https://github.com/dwave-examples/flow-shop-scheduling/pull/5/commits/6c9e8ca64844a1f71962ed15d87350bde9b8ee47

File: app.py (hunk 6 above)
Comment (k8culver) [reply app.py]: updated: https://github.com/dwave-examples/flow-shop-scheduling/pull/5/commits/6c9e8ca64844a1f71962ed15d87350bde9b8ee47

File: app.py (hunk 5 above)
Comment (k8culver) [reply app.py]: updated: https://github.com/dwave-examples/flow-shop-scheduling/pull/5/commits/6c9e8ca64844a1f71962ed15d87350bde9b8ee47

File: dash_html.py (hunk 7 above)
Comment (thisac) [reply dash_html.py]: I suspect the comma might have been there before (making it a tuple) with black just adding the parentheses , although I might be wrong. Black can be a mystery.

**Comment by k8culver** [comment]
> * Not sure if this is easily fixable, but it'd be nice to have the conflicts button switch between the two graphs as smoothly as the sort button. Can both images be rendered at start and just hidden in the same was as the sorted ones?

The reason this was happening was because there's a loading wrapper around the graph which is necessary for switching between scenarios and for first page load but makes it look awkward when switching between show/hide conflicts. I found that Dash Loading has a cool feature that lets you delay the start of the load so by setting it to 200ms it doesn't show the loading icon for short loads. Using this feature required a Dash upgrade though so I've upgraded the requirement to 2.17.1

https://github.com/dwave-examples/flow-shop-scheduling/pull/5/commits/6168ff69af2cd4e620ed759fb9bc9eb189c14862

**Comment by k8culver** [comment]
> * The legend now has a conflicts column (adding an _extra_ item for each "conflicted" colour) which I don't think is necessary. It looks better, and to me clearer, just to have the colours in the legend.

@thisac I haven't been able to find a solution for this. Do you have any ideas?

File: src/generate_charts.py [hunk 9]
```diff
@@ -153,6 +153,19 @@ def generate_gantt_chart(
...
+    # Update legend to remove conflict text and duplicates.
+    unique_labels = set()
+    for i, trace in enumerate(fig.select_traces()):
+        label = trace.legendgroup.split(",")[0]
+        if label not in unique_labels:
+            trace.update({"name": label, "legendrank": i})
+            unique_labels.add(label)
+        else:
+            fig.update_traces({"showlegend": False}, {"name": label})
+            trace.update({"name": label, "legendrank": i})
```
Comment (thisac) [inline src/generate_charts.py:165]: You could also just hide the second instance of each occurrence, unless you deliberately want those to not have the pattern in the legend, which I'm not sure makes more sense.
```suggestion
            trace.update({"showlegend": False})
```

Comment (k8culver) [reply src/generate_charts.py:165]: Yeah I don't really have a preference either way, I'm fine with this way

**Review by thisac — APPROVED** [review]
Looks fantastic! Thanks @k8culver!

**Pull request merged**